# TASK:- to predict whether a person makes over 50K per year or not from classic adult dataset using XGBoost.



# data reading and importing required packages

In [185]:
import numpy as np
import pandas as pd

train_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', header = None) #reading a training dataset.

test_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test' , skiprows = 1, header = None) #Reading a test dataset.

col_labels = ['age', 'workclass', 'fnlwgt', 'education', 'education_num',
'marital_status', 'occupation','relationship', 'race', 'sex', 'capital_gain',
'capital_loss', 'hours_per_week', 'native_country', 'wage_class'] 
train_set.columns = col_labels #assigning the collumns name
test_set.columns = col_labels  #assigning the collumns name

In [173]:
train_set.head() # look on the data set.

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  wage_class  
0          2174             0              40   United-States           0  
1             0             0              13   United-States           0  
2             0             0              40   United-States           0  
3             0             0              40   United-States           0  
4             0             0              40            Cuba           0

# data preprocesing

In [174]:
train_set.isnull().sum() #is there any null value is present? 

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
wage_class        0
dtype: int64

In [176]:
x_train=train_set.iloc[:,:-1] #deriving independent features from the data set.
x_train.corr()                #checking for correlation if present.


age    fnlwgt  education_num  capital_gain  capital_loss  \
age             1.000000 -0.076646       0.036527      0.077674      0.057775   
fnlwgt         -0.076646  1.000000      -0.043195      0.000432     -0.010252   
education_num   0.036527 -0.043195       1.000000      0.122630      0.079923   
capital_gain    0.077674  0.000432       0.122630      1.000000     -0.031615   
capital_loss    0.057775 -0.010252       0.079923     -0.031615      1.000000   
hours_per_week  0.068756 -0.018768       0.148123      0.078409      0.054256   

                hours_per_week  
age                   0.068756  
fnlwgt               -0.018768  
education_num         0.148123  
capital_gain          0.078409  
capital_loss          0.054256  
hours_per_week        1.000000

In [111]:
b=[]
for i in col: #finding the column name which has a numerical values.
    if x_train[i].dtype!=object:
        b.append(i)
b        
        
        

['age',
 'fnlwgt',
 'education_num',
 'capital_gain',
 'capital_loss',
 'hours_per_week']

In [177]:
xnew=pd.DataFrame(data=x_train[['age','fnlwgt','education_num','capital_gain','capital_loss','hours_per_week']])
xnew

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week
0       39   77516             13          2174             0              40
1       50   83311             13             0             0              13
2       38  215646              9             0             0              40
3       53  234721              7             0             0              40
4       28  338409             13             0             0              40
...    ...     ...            ...           ...           ...             ...
32556   27  257302             12             0             0              38
32557   40  154374              9             0             0              40
32558   58  151910              9             0             0              40
32559   22  201490              9             0             0              20
32560   52  287927              9         15024             0              40

[32561 rows x 6 columns]

# data scalling

In [178]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()                         #datascalling operation 
scaled_data=scaler.fit_transform(xnew)

In [179]:
xx=pd.DataFrame(data=scaled_data,columns=xnew.columns)  #conversion of scalled data into dataframe so that we can concat with encoded categorical dataframe.

In [180]:
xx

age    fnlwgt  education_num  capital_gain  capital_loss  \
0      0.030671 -1.063611       1.134739      0.148453      -0.21666   
1      0.837109 -1.008707       1.134739     -0.145920      -0.21666   
2     -0.042642  0.245079      -0.420060     -0.145920      -0.21666   
3      1.057047  0.425801      -1.197459     -0.145920      -0.21666   
4     -0.775768  1.408176       1.134739     -0.145920      -0.21666   
...         ...       ...            ...           ...           ...   
32556 -0.849080  0.639741       0.746039     -0.145920      -0.21666   
32557  0.103983 -0.335433      -0.420060     -0.145920      -0.21666   
32558  1.423610 -0.358777      -0.420060     -0.145920      -0.21666   
32559 -1.215643  0.110960      -0.420060     -0.145920      -0.21666   
32560  0.983734  0.929893      -0.420060      1.888424      -0.21666   

       hours_per_week  
0           -0.035429  
1           -2.222153  
2           -0.035429  
3           -0.035429  
4           -0.035429  
...               ...  
32556       -0.197409  
32557       -0.035429  
32558       -0.035429  
32559       -1.655225  
32560       -0.035429  

[32561 rows x 6 columns]

In [181]:
x_train.drop(x_train[['age','fnlwgt','education_num','capital_gain','capital_loss','hours_per_week']],inplace=True,axis=1) #dropping all the numeric columns. 
x_train

workclass    education       marital_status  \
0              State-gov    Bachelors        Never-married   
1       Self-emp-not-inc    Bachelors   Married-civ-spouse   
2                Private      HS-grad             Divorced   
3                Private         11th   Married-civ-spouse   
4                Private    Bachelors   Married-civ-spouse   
...                  ...          ...                  ...   
32556            Private   Assoc-acdm   Married-civ-spouse   
32557            Private      HS-grad   Married-civ-spouse   
32558            Private      HS-grad              Widowed   
32559            Private      HS-grad        Never-married   
32560       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship    race      sex  native_country  
0            Adm-clerical   Not-in-family   White     Male   United-States  
1         Exec-managerial         Husband   White     Male   United-States  
2       Handlers-cleaners   Not-in-family   White     Male   United-States  
3       Handlers-cleaners         Husband   Black     Male   United-States  
4          Prof-specialty            Wife   Black   Female            Cuba  
...                   ...             ...     ...      ...             ...  
32556        Tech-support            Wife   White   Female   United-States  
32557   Machine-op-inspct         Husband   White     Male   United-States  
32558        Adm-clerical       Unmarried   White   Female   United-States  
32559        Adm-clerical       Own-child   White     Male   United-States  
32560     Exec-managerial            Wife   White   Female   United-States  

[32561 rows x 8 columns]

In [182]:
x_new=pd.concat([xx,x_train],axis=1) # joining scalled data and categorical data 

In [183]:
x_new

age    fnlwgt  education_num  capital_gain  capital_loss  \
0      0.030671 -1.063611       1.134739      0.148453      -0.21666   
1      0.837109 -1.008707       1.134739     -0.145920      -0.21666   
2     -0.042642  0.245079      -0.420060     -0.145920      -0.21666   
3      1.057047  0.425801      -1.197459     -0.145920      -0.21666   
4     -0.775768  1.408176       1.134739     -0.145920      -0.21666   
...         ...       ...            ...           ...           ...   
32556 -0.849080  0.639741       0.746039     -0.145920      -0.21666   
32557  0.103983 -0.335433      -0.420060     -0.145920      -0.21666   
32558  1.423610 -0.358777      -0.420060     -0.145920      -0.21666   
32559 -1.215643  0.110960      -0.420060     -0.145920      -0.21666   
32560  0.983734  0.929893      -0.420060      1.888424      -0.21666   

       hours_per_week          workclass    education       marital_status  \
0           -0.035429          State-gov    Bachelors        Never-married   
1           -2.222153   Self-emp-not-inc    Bachelors   Married-civ-spouse   
2           -0.035429            Private      HS-grad             Divorced   
3           -0.035429            Private         11th   Married-civ-spouse   
4           -0.035429            Private    Bachelors   Married-civ-spouse   
...               ...                ...          ...                  ...   
32556       -0.197409            Private   Assoc-acdm   Married-civ-spouse   
32557       -0.035429            Private      HS-grad   Married-civ-spouse   
32558       -0.035429            Private      HS-grad              Widowed   
32559       -1.655225            Private      HS-grad        Never-married   
32560       -0.035429       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship    race      sex  native_country  
0            Adm-clerical   Not-in-family   White     Male   United-States  
1         Exec-managerial         Husband   White     Male   United-States  
2       Handlers-cleaners   Not-in-family   White     Male   United-States  
3       Handlers-cleaners         Husband   Black     Male   United-States  
4          Prof-specialty            Wife   Black   Female            Cuba  
...                   ...             ...     ...      ...             ...  
32556        Tech-support            Wife   White   Female   United-States  
32557   Machine-op-inspct         Husband   White     Male   United-States  
32558        Adm-clerical       Unmarried   White   Female   United-States  
32559        Adm-clerical       Own-child   White     Male   United-States  
32560     Exec-managerial            Wife   White   Female   United-States  

[32561 rows x 14 columns]

In [184]:
y_train=train_set.iloc[:,-1]  #deriving output feature or a target variable column
y_train

0        0
1        0
2        0
3        0
4        0
        ..
32556    0
32557    1
32558    0
32559    0
32560    1
Name: wage_class, Length: 32561, dtype: int64

In [ ]:
col=x_new.columns
col
x_new['age'].dtype

# one-hot encoding

In [122]:
a=[]
for i in col:                    #finding a columns which are contain categorical value
    if x_new[i].dtypes == object:
        
        a.append(i)
 
        
x_new=pd.get_dummies(data=x_new, columns=a, drop_first=True) # one hot encoding

In [159]:
x_new.head()

age    fnlwgt  education_num  capital_gain  capital_loss  \
0  0.030671 -1.063611       1.134739      0.148453      -0.21666   
1  0.837109 -1.008707       1.134739     -0.145920      -0.21666   
2 -0.042642  0.245079      -0.420060     -0.145920      -0.21666   
3  1.057047  0.425801      -1.197459     -0.145920      -0.21666   
4 -0.775768  1.408176       1.134739     -0.145920      -0.21666   

   hours_per_week  workclass_ Federal-gov  workclass_ Local-gov  \
0       -0.035429                       0                     0   
1       -2.222153                       0                     0   
2       -0.035429                       0                     0   
3       -0.035429                       0                     0   
4       -0.035429                       0                     0   

   workclass_ Never-worked  workclass_ Private  ...  native_country_ Portugal  \
0                        0                   0  ...                         0   
1                        0                   0  ...                         0   
2                        0                   1  ...                         0   
3                        0                   1  ...                         0   
4                        0                   1  ...                         0   

   native_country_ Puerto-Rico  native_country_ Scotland  \
0                            0                         0   
1                            0                         0   
2                            0                         0   
3                            0                         0   
4                            0                         0   

   native_country_ South  native_country_ Taiwan  native_country_ Thailand  \
0                      0                       0                         0   
1                      0                       0                         0   
2                      0                       0                         0   
3                      0                       0                         0   
4                      0                       0                         0   

   native_country_ Trinadad&Tobago  native_country_ United-States  \
0                                0                              1   
1                                0                              1   
2                                0                              1   
3                                0                              1   
4                                0                              0   

   native_country_ Vietnam  native_country_ Yugoslavia  
0                        0                           0  
1                        0                           0  
2                        0                           0  
3                        0                           0  
4                        0                           0  

[5 rows x 99 columns]

In [144]:
x_new.drop(x_new[["native_country_ Holand-Netherlands"]],axis=1,inplace=True) #i have dropped this columns because in given test data set there is not a single record is containing a country name  "native_country_ Holand-Netherlands".

In [143]:
x_new.columns

Index(['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week', 'workclass_ Federal-gov', 'workclass_ Local-gov',
       'workclass_ Never-worked', 'workclass_ Private',
       'workclass_ Self-emp-inc', 'workclass_ Self-emp-not-inc',
       'workclass_ State-gov', 'workclass_ Without-pay', 'education_ 11th',
       'education_ 12th', 'education_ 1st-4th', 'education_ 5th-6th',
       'education_ 7th-8th', 'education_ 9th', 'education_ Assoc-acdm',
       'education_ Assoc-voc', 'education_ Bachelors', 'education_ Doctorate',
       'education_ HS-grad', 'education_ Masters', 'education_ Preschool',
       'education_ Prof-school', 'education_ Some-college',
       'marital_status_ Married-AF-spouse',
       'marital_status_ Married-civ-spouse',
       'marital_status_ Married-spouse-absent',
       'marital_status_ Never-married', 'marital_status_ Separated',
       'marital_status_ Widowed', 'occupation_ Adm-clerical',
       'occupation_ Armed-Force

In [ ]:
l_unique = y_train.unique() # fteching out the unique values from target column
l_unique
v_unique = np.arange(len(l_unique)) # obtaining values in the range of the length of I_unique
v_unique
y_train.replace(to_replace=l_unique, value=v_unique, inplace=True) # replacing categorical data with numerical values


In [172]:
y_train.head(5)

0    0
1    0
2    0
3    0
4    0
Name: wage_class, dtype: int64


# Model Building

In [145]:
import xgboost as xgb     #importing algorithm
model=xgb.XGBClassifier(objective='binary:logistic') #object creation

In [146]:
model.fit(x_new, y_train)   #Model Training

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

# apply all the the feature engineering technichs on test data which we have applied for train data . 

In [128]:
x_test=test_set.iloc[:,:-1]

x_test

age      workclass  fnlwgt      education  education_num  \
0       25        Private  226802           11th              7   
1       38        Private   89814        HS-grad              9   
2       28      Local-gov  336951     Assoc-acdm             12   
3       44        Private  160323   Some-college             10   
4       18              ?  103497   Some-college             10   
...    ...            ...     ...            ...            ...   
16276   39        Private  215419      Bachelors             13   
16277   64              ?  321403        HS-grad              9   
16278   38        Private  374983      Bachelors             13   
16279   44        Private   83891      Bachelors             13   
16280   35   Self-emp-inc  182148      Bachelors             13   

            marital_status          occupation     relationship  \
0            Never-married   Machine-op-inspct        Own-child   
1       Married-civ-spouse     Farming-fishing          Husband   
2       Married-civ-spouse     Protective-serv          Husband   
3       Married-civ-spouse   Machine-op-inspct          Husband   
4            Never-married                   ?        Own-child   
...                    ...                 ...              ...   
16276             Divorced      Prof-specialty    Not-in-family   
16277              Widowed                   ?   Other-relative   
16278   Married-civ-spouse      Prof-specialty          Husband   
16279             Divorced        Adm-clerical        Own-child   
16280   Married-civ-spouse     Exec-managerial          Husband   

                      race      sex  capital_gain  capital_loss  \
0                    Black     Male             0             0   
1                    White     Male             0             0   
2                    White     Male             0             0   
3                    Black     Male          7688             0   
4                    White   Female             0             0   
...                    ...      ...           ...           ...   
16276                White   Female             0             0   
16277                Black     Male             0             0   
16278                White     Male             0             0   
16279   Asian-Pac-Islander     Male          5455             0   
16280                White     Male             0             0   

       hours_per_week  native_country  
0                  40   United-States  
1                  50   United-States  
2                  40   United-States  
3                  40   United-States  
4                  30   United-States  
...               ...             ...  
16276              36   United-States  
16277              40   United-States  
16278              50   United-States  
16279              40   United-States  
16280              60   United-States  

[16281 rows x 14 columns]

In [129]:
e=[]
for i in col:
    if x_test[i].dtype!=object:
        e.append(i)
e        
        
    

['age',
 'fnlwgt',
 'education_num',
 'capital_gain',
 'capital_loss',
 'hours_per_week']

In [130]:
xnewt=pd.DataFrame(data=x_test[['age','fnlwgt','education_num','capital_gain','capital_loss','hours_per_week']])
xnewt

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week
0       25  226802              7             0             0              40
1       38   89814              9             0             0              50
2       28  336951             12             0             0              40
3       44  160323             10          7688             0              40
4       18  103497             10             0             0              30
...    ...     ...            ...           ...           ...             ...
16276   39  215419             13             0             0              36
16277   64  321403              9             0             0              40
16278   38  374983             13             0             0              50
16279   44   83891             13          5455             0              40
16280   35  182148             13             0             0              60

[16281 rows x 6 columns]

In [131]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
scaled_datat=scaler.fit_transform(xnewt)

In [132]:
xxt=pd.DataFrame(data=scaled_datat,columns=xnewt.columns)
xxt

age    fnlwgt  education_num  capital_gain  capital_loss  \
0     -0.994129  0.353474      -1.196864     -0.142662     -0.218062   
1     -0.055417 -0.942391      -0.417886     -0.142662     -0.218062   
2     -0.777503  1.395450       0.750582     -0.142662     -0.218062   
3      0.377835 -0.275397      -0.028397      0.871091     -0.218062   
4     -1.499590 -0.812954      -0.028397     -0.142662     -0.218062   
...         ...       ...            ...           ...           ...   
16276  0.016791  0.245794       1.140071     -0.142662     -0.218062   
16277  1.822007  1.248371      -0.417886     -0.142662     -0.218062   
16278 -0.055417  1.755221       1.140071     -0.142662     -0.218062   
16279  0.377835 -0.998420       1.140071      0.576644     -0.218062   
16280 -0.272043 -0.068939       1.140071     -0.142662     -0.218062   

       hours_per_week  
0           -0.031432  
1            0.769918  
2           -0.031432  
3           -0.031432  
4           -0.832781  
...               ...  
16276       -0.351972  
16277       -0.031432  
16278        0.769918  
16279       -0.031432  
16280        1.571267  

[16281 rows x 6 columns]

In [133]:
y_testt=test_set.iloc[:,-1]
y_testt

0        0
1        0
2        1
3        1
4        0
        ..
16276    0
16277    0
16278    0
16279    0
16280    1
Name: wage_class, Length: 16281, dtype: int64

In [134]:
x_test.drop(x_test[['age','fnlwgt','education_num','capital_gain','capital_loss','hours_per_week']],inplace=True,axis=1)
x_test

workclass      education       marital_status          occupation  \
0            Private           11th        Never-married   Machine-op-inspct   
1            Private        HS-grad   Married-civ-spouse     Farming-fishing   
2          Local-gov     Assoc-acdm   Married-civ-spouse     Protective-serv   
3            Private   Some-college   Married-civ-spouse   Machine-op-inspct   
4                  ?   Some-college        Never-married                   ?   
...              ...            ...                  ...                 ...   
16276        Private      Bachelors             Divorced      Prof-specialty   
16277              ?        HS-grad              Widowed                   ?   
16278        Private      Bachelors   Married-civ-spouse      Prof-specialty   
16279        Private      Bachelors             Divorced        Adm-clerical   
16280   Self-emp-inc      Bachelors   Married-civ-spouse     Exec-managerial   

          relationship                 race      sex  native_country  
0            Own-child                Black     Male   United-States  
1              Husband                White     Male   United-States  
2              Husband                White     Male   United-States  
3              Husband                Black     Male   United-States  
4            Own-child                White   Female   United-States  
...                ...                  ...      ...             ...  
16276    Not-in-family                White   Female   United-States  
16277   Other-relative                Black     Male   United-States  
16278          Husband                White     Male   United-States  
16279        Own-child   Asian-Pac-Islander     Male   United-States  
16280          Husband                White     Male   United-States  

[16281 rows x 8 columns]

In [135]:
x_newt=pd.concat([xxt,x_test],axis=1)

In [84]:
x_newt

age    fnlwgt  education_num  capital_gain  capital_loss  \
0     -0.994129  0.353474      -1.196864     -0.142662     -0.218062   
1     -0.055417 -0.942391      -0.417886     -0.142662     -0.218062   
2     -0.777503  1.395450       0.750582     -0.142662     -0.218062   
3      0.377835 -0.275397      -0.028397      0.871091     -0.218062   
4     -1.499590 -0.812954      -0.028397     -0.142662     -0.218062   
...         ...       ...            ...           ...           ...   
16276  0.016791  0.245794       1.140071     -0.142662     -0.218062   
16277  1.822007  1.248371      -0.417886     -0.142662     -0.218062   
16278 -0.055417  1.755221       1.140071     -0.142662     -0.218062   
16279  0.377835 -0.998420       1.140071      0.576644     -0.218062   
16280 -0.272043 -0.068939       1.140071     -0.142662     -0.218062   

       hours_per_week      workclass      education       marital_status  \
0           -0.031432        Private           11th        Never-married   
1            0.769918        Private        HS-grad   Married-civ-spouse   
2           -0.031432      Local-gov     Assoc-acdm   Married-civ-spouse   
3           -0.031432        Private   Some-college   Married-civ-spouse   
4           -0.832781              ?   Some-college        Never-married   
...               ...            ...            ...                  ...   
16276       -0.351972        Private      Bachelors             Divorced   
16277       -0.031432              ?        HS-grad              Widowed   
16278        0.769918        Private      Bachelors   Married-civ-spouse   
16279       -0.031432        Private      Bachelors             Divorced   
16280        1.571267   Self-emp-inc      Bachelors   Married-civ-spouse   

               occupation     relationship                 race      sex  \
0       Machine-op-inspct        Own-child                Black     Male   
1         Farming-fishing          Husband                White     Male   
2         Protective-serv          Husband                White     Male   
3       Machine-op-inspct          Husband                Black     Male   
4                       ?        Own-child                White   Female   
...                   ...              ...                  ...      ...   
16276      Prof-specialty    Not-in-family                White   Female   
16277                   ?   Other-relative                Black     Male   
16278      Prof-specialty          Husband                White     Male   
16279        Adm-clerical        Own-child   Asian-Pac-Islander     Male   
16280     Exec-managerial          Husband                White     Male   

       native_country  
0       United-States  
1       United-States  
2       United-States  
3       United-States  
4       United-States  
...               ...  
16276   United-States  
16277   United-States  
16278   United-States  
16279   United-States  
16280   United-States  

[16281 rows x 14 columns]

In [138]:
k=[]
for i in x_newt.columns:
    if x_newt[i].dtypes == object:
        
        k.append(i)
 
        
x_newt=pd.get_dummies(data=x_newt, columns=k, drop_first=True)

In [139]:
x_newt.head(50)

age    fnlwgt  education_num  capital_gain  capital_loss  \
0  -0.994129  0.353474      -1.196864     -0.142662     -0.218062   
1  -0.055417 -0.942391      -0.417886     -0.142662     -0.218062   
2  -0.777503  1.395450       0.750582     -0.142662     -0.218062   
3   0.377835 -0.275397      -0.028397      0.871091     -0.218062   
4  -1.499590 -0.812954      -0.028397     -0.142662     -0.218062   
5  -0.344252  0.087571      -1.586353     -0.142662     -0.218062   
6  -0.705295  0.355593      -0.417886     -0.142662     -0.218062   
7   1.749798 -0.802274       1.919049      0.266505     -0.218062   
8  -1.066338  1.704933      -0.028397     -0.142662     -0.218062   
9   1.172129 -0.798774      -2.365331     -0.142662     -0.218062   
10  1.894216 -0.047125      -0.417886      0.703627     -0.218062   
11 -0.199834  0.217850       1.140071     -0.142662     -0.218062   
12 -0.921921 -1.015448      -0.417886     -0.142662     -0.218062   
13  1.388755  1.044306      -0.417886     -0.142662     -0.218062   
14  0.666669  0.854100      -0.417886      0.266505     -0.218062   
15  0.305626  1.482839       1.529560     -0.142662     -0.218062   
16 -1.355172  2.413341      -0.028397     -0.142662     -0.218062   
17  0.305626 -0.577814      -0.417886     -0.142662     -0.218062   
18 -0.127626 -1.219238      -0.417886     -0.142662     -0.218062   
19  0.089000 -0.987750       2.308538     -0.142662     -0.218062   
20 -0.344252 -0.771170       1.140071     -0.142662     -0.218062   
21 -0.344252  0.464966      -0.028397     -0.142662     -0.218062   
22  2.399676 -0.543182      -2.365331     -0.142662     -0.218062   
23 -0.994129  0.297936       1.140071     -0.142662     -0.218062   
24 -0.994129  0.156192       1.140071     -0.142662     -0.218062   
25  0.450043  2.302379      -0.417886      0.819665     -0.218062   
26 -1.210755  0.444524      -0.417886     -0.142662     -0.218062   
27 -1.138547 -0.520186      -0.417886     -0.142662     -0.218062   
28  1.099921 -0.850613      -0.417886     -0.142662     -0.218062   
29 -0.488669 -0.758229      -0.028397     -0.142662     -0.218062   
30  0.522252 -0.785076      -0.028397      0.871091     -0.218062   
31  1.244338 -0.026342      -1.196864     -0.142662     -0.218062   
32 -1.066338 -0.010989       1.140071     -0.142662     -0.218062   
33 -1.138547  0.649733      -0.028397     -0.142662     -0.218062   
34 -0.921921 -1.382295      -0.417886     -0.142662     -0.218062   
35  1.894216  0.022801      -0.417886     -0.142662     -0.218062   
36 -0.199834  2.026695       1.140071     -0.142662     -0.218062   
37 -1.210755  0.558220      -2.754820     -0.142662     -0.218062   
38 -1.571798  0.756722      -1.586353     -0.142662     -0.218062   
39 -1.355172  0.643953      -0.417886     -0.142662     -0.218062   
40  1.894216 -0.501853       1.529560     -0.142662     -0.218062   
41  0.377835 -0.654220       0.361093     -0.142662     -0.218062   
42 -0.199834  2.609838      -0.417886     -0.142662     -0.218062   
43 -0.705295 -0.811658      -1.196864     -0.142662     -0.218062   
44 -1.355172 -0.483056      -0.028397     -0.142662     -0.218062   
45 -0.777503  0.505113       0.361093     -0.142662     -0.218062   
46  0.016791  0.953275      -2.365331     -0.142662     -0.218062   
47  1.099921 -0.029927      -0.028397      0.372654     -0.218062   
48  0.955504  0.109981      -1.196864     -0.142662     -0.218062   
49  1.244338 -0.544119      -0.417886     -0.142662     -0.218062   

    hours_per_week  workclass_ Federal-gov  workclass_ Local-gov  \
0        -0.031432                       0                     0   
1         0.769918                       0                     0   
2        -0.031432                       0                     1   
3        -0.031432                       0                     0   
4        -0.832781                       0                     0   
5        -0.832781                       0                     0   
6        -0.0

In [140]:
from sklearn.metrics import accuracy_score

In [147]:
y_pred = model.predict(x_newt)
predictions = [round(value) for value in y_pred]
accuracy = accuracy_score(y_testt,predictions)
accuracy

0.865978748234138

In [106]:
l_unique = y_testt.unique() # fteching out the unique values from game_season/
l_unique
v_unique = np.arange(len(l_unique)) # obtaining values in the range of the length of I_unique
v_unique
y_testt.replace(to_replace=l_unique, value=v_unique, inplace=True) # 

In [150]:
y_testt.head()

0    0
1    0
2    1
3    1
4    0
Name: wage_class, dtype: int64

In [151]:
x_newt

age    fnlwgt  education_num  capital_gain  capital_loss  \
0     -0.994129  0.353474      -1.196864     -0.142662     -0.218062   
1     -0.055417 -0.942391      -0.417886     -0.142662     -0.218062   
2     -0.777503  1.395450       0.750582     -0.142662     -0.218062   
3      0.377835 -0.275397      -0.028397      0.871091     -0.218062   
4     -1.499590 -0.812954      -0.028397     -0.142662     -0.218062   
...         ...       ...            ...           ...           ...   
16276  0.016791  0.245794       1.140071     -0.142662     -0.218062   
16277  1.822007  1.248371      -0.417886     -0.142662     -0.218062   
16278 -0.055417  1.755221       1.140071     -0.142662     -0.218062   
16279  0.377835 -0.998420       1.140071      0.576644     -0.218062   
16280 -0.272043 -0.068939       1.140071     -0.142662     -0.218062   

       hours_per_week  workclass_ Federal-gov  workclass_ Local-gov  \
0           -0.031432                       0                     0   
1            0.769918                       0                     0   
2           -0.031432                       0                     1   
3           -0.031432                       0                     0   
4           -0.832781                       0                     0   
...               ...                     ...                   ...   
16276       -0.351972                       0                     0   
16277       -0.031432                       0                     0   
16278        0.769918                       0                     0   
16279       -0.031432                       0                     0   
16280        1.571267                       0                     0   

       workclass_ Never-worked  workclass_ Private  ...  \
0                            0                   1  ...   
1                            0                   1  ...   
2                            0                   0  ...   
3                            0                   1  ...   
4                            0                   0  ...   
...                        ...                 ...  ...   
16276                        0                   1  ...   
16277                        0                   0  ...   
16278                        0                   1  ...   
16279                        0                   1  ...   
16280                        0                   0  ...   

       native_country_ Portugal  native_country_ Puerto-Rico  \
0                             0                            0   
1                             0                            0   
2                             0                            0   
3                             0                            0   
4                             0                            0   
...                         ...                          ...   
16276                         0                            0   
16277                         0                            0   
16278                         0                            0   
16279                         0                            0   
16280                         0                            0   

       native_country_ Scotland  native_country_ South  \
0                             0                      0   
1                             0                      0   
2                             0                      0   
3                             0                      0   
4                             0                      0   
...                         ...                    ...   
16276                         0                      0   
16277                         0                      0   
16278                         0                      0   
16279                         0                      0   
16280                         0                      0   

       native_country_ Taiwan  native_country_ Thailand  \
0                           0                         0   
1

# Hyperparameter tunning

In [155]:
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier hyperparameter tuning

In [156]:
param_grid={
   
    ' learning_rate':[1,0.5,0.1,0.01,0.001],
    'max_depth': [3,5,10,20],
    'n_estimators':[10,50,100,200]
    
}

In [157]:
grid= GridSearchCV(XGBClassifier(objective='binary:logistic'),param_grid, verbose=3) #making a instant for gridsearchcv package

In [158]:
grid.fit(x_new,y_train) # h.p.t applied

Fitting 5 folds for each of 80 candidates, totalling 400 fits
[CV]  learning_rate=1, max_depth=3, n_estimators=10 ..................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]   learning_rate=1, max_depth=3, n_estimators=10, score=0.838, total=   1.0s
[CV]  learning_rate=1, max_depth=3, n_estimators=10 ..................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.9s remaining:    0.0s


[CV]   learning_rate=1, max_depth=3, n_estimators=10, score=0.840, total=   1.0s
[CV]  learning_rate=1, max_depth=3, n_estimators=10 ..................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.9s remaining:    0.0s


[CV]   learning_rate=1, max_depth=3, n_estimators=10, score=0.846, total=   1.0s
[CV]  learning_rate=1, max_depth=3, n_estimators=10 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=10, score=0.850, total=   1.0s
[CV]  learning_rate=1, max_depth=3, n_estimators=10 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=10, score=0.846, total=   1.0s
[CV]  learning_rate=1, max_depth=3, n_estimators=50 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=50, score=0.855, total=   4.2s
[CV]  learning_rate=1, max_depth=3, n_estimators=50 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=50, score=0.855, total=   4.2s
[CV]  learning_rate=1, max_depth=3, n_estimators=50 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=50, score=0.859, total=   4.2s
[CV]  learning_rate=1, max_depth=3, n_estimators=50 ..................
[CV]   learning_rate=1, max_depth=3, n_estimators=50, score=0.860, total=   4.2s
[CV]  l

[Parallel(n_jobs=1)]: Done 400 out of 400 | elapsed: 139.6min finished


GridSearchCV(cv=None, error_score=nan,
             estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                     colsample_bylevel=1, colsample_bynode=1,
                                     colsample_bytree=1, gamma=0,
                                     learning_rate=0.1, max_delta_step=0,
                                     max_depth=3, min_child_weight=1,
                                     missing=None, n_estimators=100, n_jobs=1,
                                     nthread=None, objective='binary:logistic',
                                     random_state=0, reg_alpha=0, reg_lambda=1,
                                     scale_pos_weight=1, seed=None, silent=None,
                                     subsample=1, verbosity=1),
             iid='deprecated', n_jobs=None,
             param_grid={' learning_rate': [1, 0.5, 0.1, 0.01, 0.001],
                         'max_depth': [3, 5, 10, 20],
                         'n_estimators': [10, 50,

In [160]:
grid.best_params_   #for finding a best perameters

{' learning_rate': 1, 'max_depth': 5, 'n_estimators': 200}

In [161]:
new_model=XGBClassifier(learning_rate= 1, max_depth= 5, n_estimators= 200)  #again training model base on hyperparameter tunning
new_model.fit(x_new, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, learning_rate=1,
              max_delta_step=0, max_depth=5, min_child_weight=1, missing=None,
              n_estimators=200, n_jobs=1, nthread=None,
              objective='binary:logistic', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, seed=None, silent=None,
              subsample=1, verbosity=1)

In [163]:
y_pred_new = new_model.predict(x_newt)                   #prediction of test data
predictions_new = [round(value) for value in y_pred_new]
accuracy_new = accuracy_score(y_testt,predictions_new)
accuracy_new

0.8407345986118788

In [169]:
y_pred_new

array([0, 0, 0, ..., 1, 0, 1], dtype=int64)

### XGBClassifier is giving me accuracy=86% without hyperparameter tunning and with hyperparameter tunning accuracy=84%